### PR TITLE
Replace `execution_mode` with `emission_type` and `boundedness`

### PR DESCRIPTION
**Pull request by jayzhan-synnada**
## Which issue does this PR close?

Closes #.

## Rationale for this change

Existing `execution_mode` is not enough the represent different combination of execution properties. There are different emission type like incrementally or emit at the end. And, we might have different decision based on the combination of boundedness of the source and emission type, for example, whether it is pipeline breaking (not able to generate result if it has unbounded input and requires to emit at the end)

In this PR we introduce `emission_type` and `boundedness` to replace the existing `execution mode` to represent more execution properties.

### CODE REVIEW

**Review by ozankabak — APPROVED** [review]
Both myself and @berkaysynnada collaborated with @jayzhan-synnada extensively on this and reviewed this API change carefully. I am quite happy with this improvement as DataFusion now exposes much more information on the nature of output streams of operators, leaving more room for downstream projects for optimizations and innovative uses.

Given that it is an API change, I think it would be great if @alamb and other interested community members can take a look and review if possible.

With this PR, we will have the following future work:
- There are some minor TODOs which will enable DataFusion to expose emission and boundedness information less conservatively for certain operators when suitable conditions exist.
- The API implicitly assumes that all output streams of an operator have the same emission and boundedness characteristics (which is the vast majority of the practical use cases). In the future, if we discover that we need to expose this information in a per-partition manner, we can extend the design to accommodate that.

File: datafusion/ffi/src/plan_properties.rs
Comment (findepi) [inline datafusion/ffi/src/plan_properties.rs:247]: Was the purpose of FFI to provide a stable ABI? cc @timsaucer

File: datafusion/common/src/join_type.rs
Comment (findepi) [inline datafusion/common/src/join_type.rs:31]: if we want to get precise, this should also say that all matches are returned (n-to-m)

File: datafusion/common/src/join_type.rs
Comment (findepi) [inline datafusion/common/src/join_type.rs:40]: > there is a match in either table

"a match" implies a matching row in left and right tables.

also  see above, this should also imply cardinality

maybe

```
/// Full Join - returns all matches between left and right table, 
/// plus left table rows without a match (with NULL values for right table columns)
/// plus right table rows without a match (with NULL values for left table columns)
```

**Comment by ozankabak** [comment]
> Thanks for expanding docstrings in JoinType. Can you please move them to separate PR? This would help with review and merging.

Makes sense, we can merge quickly ahead of the main PR

File: datafusion/ffi/src/plan_properties.rs
Comment (alamb) [reply datafusion/ffi/src/plan_properties.rs:247]: I think so too -- Maybe we can leave the `FFI_ExecutionMode` present and mark it deprecated or something 🤔

File: datafusion/physical-plan/src/execution_plan.rs
Comment (alamb) [inline datafusion/physical-plan/src/execution_plan.rs:522]: Can we clarify in these comments how Boundedness is related to the input stream?

For example, if the the input stream is unbounded, does that imply the output is also unbounded?

Maybe some examples would help to clarify this

File: datafusion/physical-plan/src/execution_plan.rs
Comment (alamb) [inline datafusion/physical-plan/src/execution_plan.rs:530]: Maybe we can give an example of each type? 

Like is it the case that
```rust
Unbounded {
  requires_infinite_memory: true,
}
```
Would describe a min/max aggregate?

File: datafusion/physical-plan/src/execution_plan.rs
Comment (alamb) [inline datafusion/physical-plan/src/execution_plan.rs:544]: I suggest documenting  when someone would expect incremental output -- I think it is basically when the operator can emit a RecordBatches of `batch_size` rows

Specifically I think it would help to say something like  "EmissionType::Incremental generates output incrementally but the operator may still buffer data internally until `batch_size` records arrive

**Review by alamb — COMMENTED** [review]
Thank you @jayzhan-synnada  and @findepi  for the reivew.

I think this PR makes a lot of sense and is well documented, reasoned, and a nice improvement. 

I have two concerns about this PR:
1. The potentially breaking changes to FFI (I am sure we can find some way around this, but I am still new to the area)
2. The accumulation of API changes we already have on main  before https://github.com/apache/datafusion/issues/13334


So maybe we can figure out the FFI stuff, and then wait to merge this PR until 44 is relased. I will try hard to focus on the 44 release later this week

File: datafusion/physical-plan/src/execution_plan.rs
Comment (berkaysynnada) [reply datafusion/physical-plan/src/execution_plan.rs:522]: Limit and TopK can be given as examples

File: datafusion/physical-plan/src/execution_plan.rs
Comment (berkaysynnada) [reply datafusion/physical-plan/src/execution_plan.rs:530]: min/max over unordered keys are `requires_infinite_memory: true` by default, but if the key cardinality is finite, then it is false. That could be added as an example.

File: datafusion/ffi/src/plan_properties.rs
Comment (timsaucer) [reply datafusion/ffi/src/plan_properties.rs:247]: Yes, the idea is to be stable. I don't think the real problem here is the change from `FFI_ExecutionMode` to `FFI_Boundness` but rather the way it breaks libraries working across the boundary that will expect one and receive the other. I think this current impact is minimal since the FFI interfaces have *just* been released and I don't think many people have had an opportunity to use them. However I will open an issue about adding a feature to validate the versioning across the boundary. I think that will be necessary going forward.

Comment (timsaucer) [reply datafusion/ffi/src/plan_properties.rs:247]: https://github.com/apache/datafusion/issues/13827

File: datafusion/physical-plan/src/execution_plan.rs
Comment (ozankabak) [reply datafusion/physical-plan/src/execution_plan.rs:522]: Sure, we can expand the comments. Basically, boundedness of the output stream is a function of the boundedness of the input stream *and* the nature of the operator. For example, a simple SISO operator that transforms it data preserves the boundedness of its input, while an operator with fetch can have an unbounded input stream but a bounded output stream.

File: datafusion/physical-plan/src/execution_plan.rs
Comment (ozankabak) [reply datafusion/physical-plan/src/execution_plan.rs:530]: Makes sense, we will add some examples. In your case, `MEDIAN` would be an example (we can do `MIN`/`MAX` with bounded memory).

**Comment by ozankabak** [comment]
> So maybe we can figure out the FFI stuff, and then wait to merge this PR until 44 is relased. I will try hard to focus on the 44 release later this week

Sure, it makes sense to wait and only proceed with merging after all the PRs that are slated for 44 get merged before this.

One alternative to consider would be making 44 a major release with all the API changes (including this) so that we can proceed with a few "stable" releases after 44. It is also not a great dev-ex to deal with some API change in every single release, albeit small.

**Review by alamb — APPROVED** [review]
Thank you @jayzhan-synnada and @ozankabak  for this PR

From my perspective this PR is fine to merge, and maybe since the FFI feature is so new as @ozankabak pointed out it would make sense to not delay this PR until after the release to get it in. 

I don't have a strong preference, though hope to have dug out of my backlog later today and help to get a release candidate in shape over the next few days

**Comment by jayzhan-synnada** [comment]
If we can merge this soon, I think we don't need to separate Join doc, just get in together. Thanks all for the review!

**Comment by ozankabak** [comment]
Great, then let's wait for a few hours in case any other feedback arrives and then merge this. This way, we will avoid API churn on FFI between DataFusion 44 and 45.

**Comment by alamb** [comment]
🚀

**Comment by alamb** [comment]
> One alternative to consider would be making 44 a major release with all the API changes (including this) so that we can proceed with a few "stable" releases after 44. It is also not a great dev-ex to deal with some API change in every single release, albeit small.

Indeed -- it may be time to consider making the API more stable. Here is a related discussion:
- https://github.com/apache/datafusion/issues/5269

**Pull request merged**
